Source/WTF/wtf/threads/Signals.cpp

```diff
@@ -203,6 +206,157 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+    // If we wanted to distinguish between SIGBUS and SIGSEGV for EXC_BAD_ACCESS on Darwin we could do:
+    // if (exceptionData[0] == KERN_INVALID_ADDRESS)
+    //    signal = SIGSEGV;
+    // else
+    //    signal = SIGBUS;
+    Signal signal = fromMachException(exceptionType);
+    RELEASE_ASSERT(signal != Signal::Unknown);
+
+    thread = threadIDGetThreadPort(threadID);
+
+    mach_msg_type_number_t stateCount = 0;
+
+#if CPU(X86_64) || CPU(X86)
+    x86_thread_state_t state;
+    int flavor = x86_THREAD_STATE;
+    stateCount = x86_UNIFIED_THREAD_STATE_COUNT;
+#elif CPU(ARM64) || CPU(ARM)
+    arm_unified_thread_state state;
+    int flavor = ARM_THREAD_STATE;
+    stateCount = ARM_UNIFIED_THREAD_STATE_COUNT;
+#endif
+
+    kr = thread_get_state(thread, flavor, (thread_state_t)&state, &stateCount);
+    if (kr != KERN_SUCCESS) {
+        dataLogLn("thread_get_state failed due to ", mach_error_string(kr));
+        return kr;
+    }
+
+#if CPU(ARM64E) && OS(DARWIN)
+    ptrauth_generic_signature_t inStateHash = hashThreadState((thread_state_t)&state);
+#endif
+
+#if CPU(X86_64)
+    PlatformRegisters& registers = state.uts.ts64;
+    ASSERT(state.ash.flavor == X86_THREAD_STATE64);
+#elif CPU(X86)
+    PlatformRegisters& registers = state.uts.ts32;
+    ASSERT(state.ash.flavor == X86_THREAD_STATE32);
+#elif CPU(ARM64)
+    PlatformRegisters& registers = state.ts_64;
+    ASSERT(state.ash.flavor == ARM_THREAD_STATE64);
+#elif CPU(ARM)
+    PlatformRegisters& registers = state.ts_32;
+    ASSERT(state.ash.flavor == ARM_THREAD_STATE32);
+#endif
+
+    SigInfo info;
+    if (signal == Signal::AccessFault) {
+        ASSERT_UNUSED(dataCount, dataCount == 2);
+        info.faultingAddress = reinterpret_cast<void*>(exceptionData[1]);
+#if CPU(ADDRESS64)
+        // If the faulting address is out of the range of any valid memory, we would
+        // not have any reason to handle it. Just let the default handler take care of it.
+        static constexpr unsigned validAddressBits = OS_CONSTANT(EFFECTIVE_ADDRESS_WIDTH);
+        static constexpr uintptr_t invalidAddressMask = ~((1ull << validAddressBits) - 1);
+        if (bitwise_cast<uintptr_t>(info.faultingAddress) & invalidAddressMask)
+            return KERN_FAILURE;
+#endif
+    }
+
+    bool didHandle = false;
+    handlers.forEachHandler(signal, [&] (const SignalHandler& handler) {
+        SignalAction handlerResult = handler(signal, info, registers);
+        didHandle |= handlerResult == SignalAction::Handled;
+    });
+
+    if (didHandle) {
+#if CPU(ARM64E) && OS(DARWIN)
+        RELEASE_ASSERT(inStateHash == hashThreadState((thread_state_t)&state));
+#endif
+        kr = thread_set_state(thread, flavor, (thread_state_t)&state,
+            stateCount);
+
+        if (kr != KERN_SUCCESS)
+            dataLogLn("thread_set_state failed due to ", mach_error_string(kr));
+
+        return kr;
+    }
+    return KERN_FAILURE;
+}
```

Comment:
This logic is shared with the implementation of `catch_mach_exception_raise_state` we should split into a helper function.

Reply:
I don't actually think it's a win here, but I'll defer to your advice.